### ISSUE TITLE
Isn't s_binary() supposed to be fuzzable?

### ISSUE THREAD

**Issue by kuspilur** [jtpereyda/boofuzz#131]
According to the documentation s_binary() is fuzzable, however it doesn't seems to support the fuzzable=True/False keyword and is not fuzzed during execution.

If it is not fuzzable then how can I represent and fuzz a binary or hex value that I extract directly from a pcap file with network communication?

**Comment by cbisnett** [comment]
I think you might be confused. I looked at the Sulley manual (http://www.fuzzing.org/wp-content/SulleyManual.pdf) and I see these two snippets:
> Let's begin with the simplest primitive, s_static(), which adds a static unmutating value of arbitrary length to the request. There are various aliases sprinkled throughout Sulley for your convenience, s_dunno(), s_raw() and s_unknown() are aliases of s_static():
>...
>Related to the above, but not equivalent, is the s_binary() primitive which accepts binary data represented in multiple formats.

The second quote is states that `s_binary` is related to the above because it's static unmutated data but it's not equivalent because it parses the value to translate from different formats into a string of binary data.

**Comment by jtpereyda** [comment]
I had to go check the source, but the documentation @cbisnett referenced has a nice example:
```python
# yeah, it can handle all these formats.
s_binary("0xde 0xad be ef \xca fe 00 01 02 0xba0xdd f0 0d", name="complex")
```
So as @cbisnett said, it basically translates a hex string into a static value. I don't think there's any reason this one couldn't be fuzzable, it just hasn't been implemented yet.

I'll consider this a feature request for fuzzing binary chunks.

If you have PCAP data though, you can also reconstruct the protocol format using smaller primitives. See https://github.com/jtpereyda/boofuzz-ftp for a fairly simple example.

**Comment by jtpereyda** [comment]
It would be nice to take raw pcap data and fuzz from that; I wonder if there's a tool for that use case out there... seems like there should be. (wouldn't be as "smart" as describing the protocol in boofuzz)

**Comment by bsmelo** [comment]
Sorry to drop in the discussion like this and all, but wouldn't "take raw pcap data and fuzz from that" fall into Mutational Fuzzing instead of Generational? (not that this is a problem at all)

Why I say this: I'm in a situation where I want to develop a Generational **and** a Mutational fuzzer for a given protocol, and my initial thought was to just write two different tools: the gen_fuzzer based on boofuzz, and a mut_fuzzer based on scapy (which has functions to deal with pcap files and to read those pcap files - raw hex dumps - into scapy layers, split by protocols fields, etc).

**Comment by cbisnett** [comment]
One of the key features of Sulley and Boofuzz is that the format contains default values. The point of the default value is to provide a value that is valid for the format and will allow the target to continue processing past the field.

So I think the idea here is to allow users to carve "valid data" from a PCAP to provide as the default value. Then as they continue to reverse engineer or learn the protocol, then the blob of data can be broken up into fields with default values. This would shorten some of the initial work required to develop a generative fuzzer.

**Comment by jtpereyda** [comment]
@bsmelo Correct. If you're going that direction now, using two separate tools is probably the best way.

@cbisnett I know Defensics sometimes lets you give PCAPs as input to their generational fuzz tool, but I'm not sure what exactly it does with the data.
